### DISCUSSION TITLE
Unclear migration path: revalidate => mutate

### DISCUSSION THREAD

**Discussion by konradjurk** [vercel/swr#1716] [Q&A]
Hello,

we are trying to migrate from `^0.3.5` to `^1.1.1`. [According to the migration guide](https://swr.vercel.app/blog/swr-v1#change-revalidate-to-mutate), it should be possible to replace all `revalidate()` calls with `mutate()`. However, mutate does not seem to update the `data`, in contrast to `revalidate()`. 

0.3.5 working:

```
const settingsSWR = useSWR<Settings>(...);

console.log('beforeUpdate', settingsSWR.data)

await settingsSWR.mutate(settingsPayload, false);

await settingsSWR.revalidate();

console.log('afterUpdate', settingsSWR.data)
```

Now, the old data is logged once as "beforeUpdate" and the updated data as "afterUpdate"

1.1.1 not working

```
const settingsSWR = useSWR<Settings>(...);

console.log('beforeUpdate', settingsSWR.data)

await settingsSWR.mutate(settingsPayload, false);

await settingsSWR.mutate();

console.log('afterUpdate', settingsSWR.data)
```

Now, both "beforeUpdate" and "afterUpdate" log the old data.

What am I doing wrong?

Thank you!

**Comment by shuding** [comment]
This might be a bug, I'll take a look shortly. Thanks for reporting :)

**Comment by shuding** [comment]
Hi! After looking deeper into this, it's a more or less expected behavior. Because `settingsSWR` here is a __snapshot__ value of the state, and it shouldn't change over time. This is just like React's `props`:

```js
function Button(props) {
  return <button onClick={() => {
    setTimeout(() => {
      alert('You clicked ' + props.text + '!')
    }, 1000)
  }}>{props.text}</button>
}
```

If you click the button, and re-render the component with another `text` quickly, the alert will still show the previous `text` because `props` is a snapshotted value, not a reference of the latest one.

It should work the same in SWR (I would say that the old behavior is a bug):

```js
console.log(settingsSWR.data)
await anyAsynchronousAction()
console.log(settingsSWR.data) // always the same
```

The snapshot should be immutable.

To correctly implement the reference value, you will have to use `useRef` in React:

```js
const settingsSWR = useSWR(...)
const latestData = useRef()
useEffect(() => {
  latestData.current = settingsSWR.data
}, [settingsSWR.data])

// ...
console.log(latestData.current)
await anyAsynchronousAction()
console.log(latestData.current) // → updated!
```
